### ISSUE TITLE
SF Invalid geometries: Linestring - selfintersecting

### ISSUE THREAD

**Issue by chris-english** [edzer/sfr#174]
Probably an enhancement request. I had rather hoped that 'self-intersecting' line strings would be okay in the sf view as it is generalizable to many 'where did they go' kinds of questions involving 2D spaces where the actor of interest recrosses its prior path. I can understand the difficulty if linestring is viewed as precursor (or merely precursor) to polygon, but there are many domains where crossing of a path is just to be expected, not just in traveling salesmen (where its probably not allowed, but in walking my dogs, predators, pizza delivery.) 

And I suppose it could be a different 'linestring' if M is employed and M != M is part of the intersection test (i.e. not crossing due to difference in whatever metric M is in), but I'm thinking the test is gIntersect in rgeos which might not bother to determine a 'difference in M, in the metric of M' prior to declaring error, or it is a three position windowing restriction; over three without selfintersection and you're ok. I've thus far built a linestring of length 1400+ with clear crossings, but not necessarily withing a 3 window .

So might we hope for a gIntersect = FALSE/TRUE for those of us who study those who trip over their own feet, or cross their own paths? I ask as I've also wondered if there might be an GEOS way to reformulate linestring(s) to multilinestring (with tolerance before and after) at points of self-intersection to both preserve geometric validity and validity across attributes ("This is the same path as it was before, what the same attributes, but it, unfortunately, crossed a prior iteration of itself...And so the 'sameness' aspect is a little difficult in the presence of the intersect test.) both before and after intersection, essentially to pass the intersection and call the data before and after the 'to be ignored' intersection the same data until the next linestring point. 

I am sure there are many good reasons for non-selfintersecting linestrings (multilinestrings) in many domains, but I sense there are as many countervailing reasons in other domains and it might be handy to have access to the switching the test of intersection on or off depending.

**Comment by edzer** [comment]
I think what you're after is a representation of a trajectory, which is not well represented by a linestring. Did you look at [this](https://gis.stackexchange.com/questions/16347/is-there-a-postgis-function-for-determining-whether-a-linestring-intersects-itse) SO answer? We have UnaryUnion when `st_union` is called with a single argument:
```
> st_as_text(st_union(st_linestring(rbind(c(0,0),c(1,1),c(0,1),c(1,0)))))
[1] "MULTILINESTRING((0 0, 0.5 0.5), (0.5 0.5, 1 1, 0 1, 0.5 0.5), (0.5 0.5, 1 0))"
```

**Comment by mdsumner** [comment]
@chris-english I understand what you're after generally,  and as Edzer says  it's really not what simple features is for, you can get by to a certain degree but you end up hitting the walls and things get complicated. You can put more attributes on the coordinates, but ultimately you can't index a duplicated vertex in a nested object, without having that reference another table or array. TopoJSON does that, to a certain degree, it nests indexes using the simple features model, but the coordinates are necessarily in separate arrays, and it is purely planar, no Z or time. 


Some of the things you ask about are more possible by decomposing to individual line segments, and *you can* use simple features like that - you can relatively easily treat sets  of simple features in very flexible ways, in ways you never could with sp - perhaps you could provide a specific example of what you're trying to solve and we might reformulate it?  One thing you'll miss for now (I think ?) is the raw speed of `sp::over` which had all the smart bbox-test optimizations for many-to-many tests, but then sf is much faster in other ways, and fast enough to translate to sp to drive over, and then back again. You might end up with multiple geometry columns or collections of data frames to keep the indexes making sense, but it's been surprising to me how much you can do flexibly, and the prospects for leveraging other tools by translating in and out are pretty good.

**Comment by edzer** [comment]
@mdsumner - regarding speed, there's still the potential of prepared geometries in GEOS - they're in rgeos, but not yet used in sf. Essentially a spatial index built on-the-fly, and discarded.

**Comment by mdsumner** [comment]
Hmm that indexing would be very handy in long form, that's how I drive cached extraction with raster, df of feature, cell rather than lists of vectors. Then there's one extract(r, cell) call for each incoming 2D slice, and a clear target structure. Maybe there's a general approach there for IDs in sf, at least as an alternative to the default.

**Comment by chris-english** [comment]
@edzer, @mdsumner:  UnaryUnion (forehead slap), line segmentize (forehead slap). The more I go into R the more I find myself back in POSTgis.  Thank you both for your patience.

**Comment by michaeldgarber** [comment]
I had a similar challenge with self-intersecting geometries. Sharing my solution (completely in sf) in case helpful.

I needed to determine what proportion of each bike route (MULTILINESTRING), overlapped a given type of infrastructure (e.g. bike lane), also MULTILINESTRING. Many of the bike routes are self-intersecting, e.g.  a commute out and back, and I needed to preserve the self-intersection to reflect the true route length.

On my first try, I created an `st_buffer `around the infrastructure and found the `st_intersection` between the two sf objects. As expected, the self-intersecting part of the route was only counted once in the `st_intersection`, confirmed with `st_length` before and after.

To solve, I converted the "sfc_MULTILINESTRING" object into "sfc_LINESTRING" using `st_cast(x, "LINESTRING")`. (I had tried `st_segmentize` to chop into smaller pieces thereafter but wasn't having luck.) This dis-aggregated the data from ~ 100 rows, each of MULTILINESTRING to ~ 1,000 rows of LINESTRING. I then performed a [unary union ](https://r-spatial.github.io/sf/reference/geos_combine.html) with `st_union(x)` and measured `st_length` before and after. Those half as long as original after `st_union(x)` were deemed self-intersecting (max, twice in this case), which I flagged. 

I then found the `st_intersection` of the linestring-level object with the buffered bike infrastructure layer, by feature. This further dis-aggregated each linestring, i.e many rows for each linestring. For each linestring segment whose parent had been self-intersecting (but again was only counted once in the `st_intersection`), I multiplied its length by two. I then summarized the length back to the original linestring level then to the route level using aspatial operations (group by(), summarize(), left_join(), etc.).
